app/publish/publish.go

```diff
@@ -36,14 +39,35 @@ type LbrynetPublisher struct {
 // UploadHandler glues HTTP uploads to the Publisher.
 type UploadHandler struct {
 	Publisher  Publisher
-	uploadPath string
+	UploadPath string
+}
+
+type UploadOpts struct {
+	Path         string
+	Publisher    Publisher
+	ProxyService *proxy.Service
 }
 
 // NewUploadHandler returns a HTTP upload handler object.
-func NewUploadHandler(uploadPath string, publisher Publisher) UploadHandler {
+func NewUploadHandler(opts UploadOpts) UploadHandler {
+	var (
+		publisher  Publisher
+		uploadPath string
+	)
+	if opts.ProxyService != nil {
+		publisher = &LbrynetPublisher{Service: opts.ProxyService}
+	} else if opts.Publisher != nil {
```

Comment:
what if `ProxyService` and `opts.Publisher` are nil? then you are return an empty publisher with handler. Does this matter?

Reply:
In this case it will be initialized with an empty publisher and that will bite in the rear when someone tries to publish a file with this instance. It's not a good idea to set a default there.

I thought I'd rather let it fail in case of a programmer's error like this but then I found examples of constructors returning  `*Val, error` in the stdlib so I'll fix it to do that.